libs/platform/ebpf_hash_table.c

```diff
@@ -747,80 +778,93 @@ ebpf_hash_table_delete(_Inout_ ebpf_hash_table_t* hash_table, _In_ const uint8_t
+ebpf_hash_table_iterate(
     _In_ const ebpf_hash_table_t* hash_table,
-    _In_opt_ const uint8_t* previous_key,
-    _Outptr_ uint8_t** next_key_pointer,
-    _Outptr_opt_ uint8_t** value)
+    _Inout_ uint64_t* cookie,
+    _Outptr_ const uint8_t** next_key,
+    _Outptr_opt_ uint8_t** next_value)
 {
-    ebpf_result_t result = EBPF_SUCCESS;
-    uint32_t hash;
-    ebpf_hash_bucket_entry_t* next_entry = NULL;
-    size_t bucket_index;
-    size_t data_index;
-    bool found_entry = false;
+    uint32_t bucket = 0;
+    uint32_t data_index = 0;
+    ebpf_result_t result;
+    if (!hash_table || !cookie || !next_key) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    bucket = (uint32_t)(*cookie >> 32);
+    data_index = (uint32_t)(*cookie & UINT32_MAX);
 
-    if (!hash_table || !next_key_pointer) {
+    if (bucket >= hash_table->bucket_count) {
         result = EBPF_INVALID_ARGUMENT;
         goto Done;
     }
 
-    hash = (previous_key != NULL) ? _ebpf_hash_table_compute_hash(hash_table, previous_key) : 0;
+    *next_key = NULL;
+    if (next_value) {
+        *next_value = NULL;
+    }
+
+    for (; bucket < hash_table->bucket_count; bucket++) {
+        ebpf_hash_bucket_header_t* bucket_header = hash_table->buckets[bucket].header;
 
-    for (bucket_index = hash % hash_table->bucket_count; bucket_index < hash_table->bucket_count; bucket_index++) {
-        ebpf_hash_bucket_header_t* bucket = hash_table->buckets[bucket_index].header;
         // Skip empty buckets.
-        if (!bucket) {
+        if (!bucket_header) {
             continue;
         }
 
-        // Pick first entry if no previous key.
-        if (!previous_key) {
-            next_entry = _ebpf_hash_table_bucket_entry(hash_table->key_size, bucket, 0);
-            break;
+        // If we are at the end of the current bucket, visit the next bucket.
+        if (data_index >= bucket_header->count) {
+            data_index = 0;
+            continue;
         }
 
-        for (data_index = 0; data_index < bucket->count; data_index++) {
-            ebpf_hash_bucket_entry_t* entry = _ebpf_hash_table_bucket_entry(hash_table->key_size, bucket, data_index);
-            if (!entry) {
-                result = EBPF_INVALID_ARGUMENT;
-                goto Done;
-            }
-            // Do we have the previous key?
-            if (found_entry) {
-                // Yes, then this is the next key.
-                next_entry = entry;
-                break;
-            }
-
-            // Is this the previous key?
-            if (_ebpf_hash_table_compare(hash_table, previous_key, entry->key) == 0) {
-                // Yes, record its location.
-                found_entry = true;
-            }
-        }
-        if (next_entry) {
-            break;
+        ebpf_hash_bucket_entry_t* entry =
+            _ebpf_hash_table_bucket_entry(hash_table->key_size, bucket_header, data_index);
+        data_index++;
```

Comment:
an index isn't robust to deletions, is it?  is there something that protects this?